Model Value, Asset and Nonce confidential variants using secp256k1-zkp types

luckysori opened the pull request:
This is the part of PR https://github.com/ElementsProject/rust-elements/pull/70 which should be less controversial.

When splitting up that PR and going through @sanket1729's review I noticed that we had broken the serde implementations of `Value`, `Asset` and `Nonce` (and I found [this](https://github.com/ElementsProject/rust-secp256k1-zkp/pull/10) (our) bug in rust-secp256k1-zkp). 

This also made me realise that, by delegating to said library when implementing serde on our types, we had introduced breaking changes. I've documented the original serde implementations with the tests in https://github.com/ElementsProject/rust-elements/commit/1c53b177c3e56c21836fa3f89bb4b4f2da0689ef, to be able to highlight the changes introduced in https://github.com/ElementsProject/rust-elements/commit/516ac0ee35ed5754ab9e8d76f3cfb60643b51a4b. I made some choices there which could be questionable, so please take a look.

---

Once https://github.com/ElementsProject/rust-secp256k1-zkp/pull/10 is merged, we should remove the patch dependency to `luckysori/rust-secp256k1-zkp` in `Cargo.toml`.

thomaseizinger commented on src/confidential.rs:
```diff
@@ -419,22 +741,29 @@ mod tests {
-                Token::U8(1), Token::U8(1), Token::U8(1), Token::U8(1),
-                Token::U8(1), Token::U8(1), Token::U8(1), Token::U8(1),
-                Token::U8(1), Token::U8(1), Token::U8(1), Token::U8(1),
-                Token::TupleEnd,
```
So this used to serialize as a tuple? Very interesting! I guess for a serializer impl like JSON, it doesn't make a difference because JSON doesn't have support for tuples?

thomaseizinger commented on src/encode.rs:
```diff
@@ -53,9 +58,17 @@ impl fmt::Display for Error {
                 requested: ref r,
                 max: ref m,
-            } => write!(f, "oversized vector allocation: requested {}, maximum {}", r, m),
+            } => write!(
+                f,
+                "oversized vector allocation: requested {}, maximum {}",
+                r, m
+            ),
```
Nit: This is just a format change, not sure if it is wanted / tolerated?

thomaseizinger commented on src/transaction.rs:
```diff
@@ -1661,7 +1660,7 @@ mod tests {
         // Output with pushes that are e.g. OP_1 are nulldata but not pegouts
         let output: TxOut = hex_deserialize!("\
-            0a2d3634393536d9a2d0aaba3823f442fb24363831fdfd0101010101010101010\
+            0a319c0000000000d3d3d3d3d3d3d3d3d3d3d3d3fdfdfd0101010101010101010\
```
Do you remember why this changed?

sanket1729 replied:
The test was incorrect. The asset is not a correct curve point

thomaseizinger commented on src/transaction.rs:
```diff
@@ -373,10 +375,165 @@ impl Decodable for TxOut {
+pub enum TxOutError {
+    NoBlindingKeyInAddress,
+    Upstream(secp256k1_zkp::Error),
+}
```
This should implement `std::error::Error`! :)

thomaseizinger commented on src/transaction.rs:
```diff
@@ -376,13 +376,35 @@ impl Decodable for TxOut {
     Upstream(secp256k1_zkp::Error),
 }
 
+impl fmt::Display for ConfidentialTxOutError {
```
Can we squash this into the one where it is introduced?

thomaseizinger commented on src/transaction.rs:
```diff
@@ -404,6 +404,11 @@ impl std::error::Error for ConfidentialTxOutError {
     }
 }
 
+impl From<secp256k1_zkp::Error> for ConfidentialTxOutError {
```
Also a candidate for squashing I'd say.

thomaseizinger commented on src/hash_types.rs:
```diff
@@ -55,9 +52,3 @@ impl_hashencode!(Wtxid);
 impl_hashencode!(BlockHash);
 impl_hashencode!(TxMerkleNode);
-
-impl ThirtyTwoByteHash for SigHash {
```
If this is deleted here, we should just drop the previous commit.

luckysori replied:
I must have messed something up, because this change was desired. I'll fix it.

thomaseizinger commented on src/script.rs:
```diff
@@ -1158,8 +1158,6 @@ mod test {
         assert!(script_2 < script_3);
         assert!(script_3 < script_4);
 
-        assert_eq!(script_1, script_1);
```
:grin:

thomaseizinger reviewed (commented):
Some nits, mainly concerning patch history :)

Thanks for pushing the upstream forward!

sanket1729 commented on src/script.rs:
```diff
@@ -1147,7 +1147,7 @@ mod test {
     }
 
-	#[test]
+        #[test]
```
indent

luckysori replied:
This is bugging me because locally I don't have this change D:

thomaseizinger replied:
Make sure you are diffing against current master. Maybe another rebase helps!

sanket1729 commented on src/script.rs:
```diff
@@ -1158,8 +1158,7 @@ mod test {
         assert!(script_2 < script_3);
         assert!(script_3 < script_4);
 
-        assert!(script_1 <= script_1);
```
These lines should not be deleted

luckysori replied:
Just curious, what's this supposed to catch? Doesn't the compiler enforce that `script_1` is equal to `script_1`? I've replaced it with `assert_eq`, but maybe you actually wanted the original assertions. Let me know.

sanket1729 replied:
I think the test is script is for `script_ord` which is supposed to check basic comparison orderings like `script_1 >= script_1`. Even if it tests something that is trivial, it should not be removed as a part of this PR.

luckysori replied:
I've reverted it back to the original state.

sanket1729 reviewed (commented):
Overall, I think this looks good, left some minor code style nits. It would be great to also incorporate suggestions which Thomas Eizenger suggested.

I am not sure about the `serde` breaking change. I will let @apoelstra comment on it.

luckysori commented:
Squashed and I tried to deal with all the nits. Maybe someone else can deal with the silly indent in `script.rs` line 1150, because it's not showing up on my machine :shrug:

apoelstra commented:
Sorry for the long delay in reviewing.

concept ACK everything, including the breaking serde changes. (I'm not thrilled about using a sequence of (tag, bytes) like this, but I like my original scheme even less .... would prefer if we serialized as a flat sequence of bytes or a hex string, similar to what `PublicKey` et al do).

There are two problems blocking this now:

1. The unit tests fail with `cargo test --features="serde-feature"` as of 1c53b177c3e56c21836fa3f89bb4b4f2da0689ef
2. The cargo patch is still in place as of 285f5656 ... and this doesn't compile anymore since the branch has been deleted!

Unsure why CI is not catching the unit test failure :( does anyone else see this?

sanket1729 commented:
@apoelstra , I think both of those are related to deleted branch. I have the branch saved locally, so all commits in the PR compile and test with `serde-feature` correctly. 

The branch was probably not deleted when the CI ran, so the CI did not detect the failure

sanket1729 commented:
Looks like the branch was merged into rust-secp-zkp master, that is why it was deleted.

sanket1729 commented:
rust-secp-zkp `0.2.1` has been released. So, we don't need any patch dependencies :)

apoelstra commented:
Confirmed that a modified version of this PR with the `[patch]` removed now passes (unit tests pass on every commit)

apoelstra commented:
Sorta-ack 7a248e0e7f8b464890e3137b62011524e394a237

This PR looks great! Thanks so much for this work, especially doing the fiddly computations for last vs not-last `TxOut`s. If you can remove the `[patch]` from Cargo.toml I'm happy to merge this.

luckysori commented:
@apoelstra 

> Sorta-ack [7a248e0](https://github.com/ElementsProject/rust-elements/commit/7a248e0e7f8b464890e3137b62011524e394a237)

I actually noticed that I had set `RANGEPROOF_MIN_VALUE` to 0 instead of 1. Maybe it's not a problem, but I've changed it back to 1 (see 38e89fb).

> This PR looks great! Thanks so much for this work, especially doing the fiddly computations for last vs not-last `TxOut`s. If you can remove the `[patch]` from Cargo.toml I'm happy to merge this.

Thank you! I've now removed the patch dependency from all commits.
